src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.cs

```diff
@@ -0,0 +1,223 @@
+        // their respective tables.
+
+        private readonly SolutionAccessor _solutionAccessor;
+        private readonly ProjectAccessor _projectAccessor;
+        private readonly DocumentAccessor _documentAccessor;
+
+        // We pool connections to the DB so that we don't have to take the hit of 
+        // reconnecting.  The connections also cache the prepared statements used
+        // to get/set data from the db.  A connection is safe to use by one thread
+        // at a time, but is not safe for simultaneous use by multiple threads.
+        private readonly object _connectionGate = new object();
+        private readonly Stack<SqlConnection> _connectionsPool = new Stack<SqlConnection>();
+
+        public SQLitePersistentStorage(
+            IOptionService optionService,
+            string workingFolderPath,
+            string solutionFilePath,
+            string databaseFile,
+            Action<AbstractPersistentStorage> disposer)
+            : base(optionService, workingFolderPath, solutionFilePath, databaseFile, disposer)
+        {
+            _solutionAccessor = new SolutionAccessor(this);
+            _projectAccessor = new ProjectAccessor(this);
+            _documentAccessor = new DocumentAccessor(this);
+        }
+
+        private SqlConnection GetConnection()
+        {
+            lock (_connectionGate)
+            {
+                // If we have an available connection, just return that.
+                if (_connectionsPool.Count > 0)
+                {
+                    return _connectionsPool.Pop();
+                }
+            }
+
+            // Otherwise create a new connection.
+            return new SqlConnection(this.DatabaseFile);
+        }
+
+        private void ReleaseConnection(SqlConnection connection)
+        {
+            lock (_connectionGate)
+            {
+                // If we've been asked to shutdown, then don't actually add the connection back to 
+                // the pool.  Instead, just close it as we no longer need it.
+                if (_shutdownTokenSource.IsCancellationRequested)
+                {
+                    connection.Close_OnlyForUseBySqlPersistentStorage();
+                    return;
+                }
+
+                _connectionsPool.Push(connection);
+            }
+        }
+
+        public override void Close()
+        {
+            lock (_connectionGate)
+            {
+                // Notify any outstanding async work that it should stop.
+                _shutdownTokenSource.Cancel();
+
+                // Go through all our pooled connections and close them.
+                while (_connectionsPool.Count > 0)
+                {
+                    var connection = _connectionsPool.Pop();
+                    connection.Close_OnlyForUseBySqlPersistentStorage();
+                }
+            }
+        }
+
+        private PooledConnection GetPooledConnection()
```

Comment:
I don't really like that this is private but accessed by the other nested types.